### ISSUE TITLE
Reset or shrink momentum vector when shrinking learning rate

### ISSUE THREAD

**Issue by reunanen** [davisking/dlib#869]
Not sure at all, but to me it looks like any momentum is preserved after learning rate is shrunk.

If so, then I guess as a result the optimizer might end up going in an unfavorable direction for quite a while, which may take a very long time to backtrack, given the new (smaller-than-before) learning rate.

I think it would make sense to reset or shrink also the momentum vector, whenever the learning rate is shrunk.

@davisking, can you comment if the above makes any sense? Just an observation from a single training run – didn't look in a lot of detail yet. If it already works like this, then I'm happy to close the issue right away.

**Comment by reunanen** [comment]
I mean, even getting the momentum to zero may take quite a while if `learning_rate_shrink` is small.

**Comment by davisking** [comment]
Do other frameworks do this?  In general I think they don't and many SGD
algorithms definitely don't do that kind of thing.  You can try it and see
what happens.  But I expect it's not better.

**Comment by reunanen** [comment]
I have no idea if other frameworks do this. Just thought it'd make sense. Will perhaps try at some point.

**Comment by reunanen** [comment]
I guess it makes less difference if learning rate is decreased gradually and not quickly by a factor of 10 or so.

**Comment by davisking** [comment]
Maybe.

**Comment by reunanen** [comment]
In fact, I observed the following sequence of events:

1. Loss goes up for a while.
2. As a consequence, learning rate is shrunk.
3. Loss keeps going up, possibly due to momentum.
4. Learning rate is shrunk even more, until it's really really small.
5. Loss keeps growing at a speed that can't be explained by the new, very small learning rate. Guessing it had to be momentum.

**Comment by davisking** [comment]
It could be, but the momentum term gets small fast. With a momentum of 0.9
it takes about 22 steps to shrink it by a factor of 10, which is not very
many steps.   The "steps without progress" metric is going to be set to
look at way more than 20 steps when it's deciding if it should shrink the
learning rate.  So this shouldn't matter.

**Comment by reunanen** [comment]
Right. Maybe it was something else then. Closing the issue, at least for now.

**Comment by reunanen** [comment]
Indeed the behaviour is (qualitatively) the same even with momentum = 0. So it must be something else.

**Comment by davisking** [comment]
There is also backtracking built into dnn_trainer.  If it sees that the loss suddenly goes up it should reload the solver state from disk and try again.  That should happen unless you have disabled it, assuming the loss is really going up.  Unless it's going up really slowly.

**Comment by reunanen** [comment]
Right. In fact reloading from disk was not happening, apparently because in my case [this](https://github.com/davisking/dlib/blob/c2a9dee846cad7b20998ee0f3fae79b6bcb67d43/dlib/dnn/trainer.h#L1067) is always true. That seems to be a result of me calling `get_net` explicitly during training (each 20 steps actually... so `gradient_updates_since_last_sync` is always less than 30).

Will clearly need to change something.

**Comment by davisking** [comment]
Calling `get_net` shouldn't have anything to do with this.  That variable
is counting how many updates since it last saved to disk.  The problem is
either that you configured it to save to disk way too often or the steps
without progress threshold is set way too low.

**Comment by reunanen** [comment]
`get_net` [calls](https://github.com/davisking/dlib/blob/c2a9dee846cad7b20998ee0f3fae79b6bcb67d43/dlib/dnn/trainer.h#L141) `sync_to_disk(true);`, which in turn [sets](https://github.com/davisking/dlib/blob/c2a9dee846cad7b20998ee0f3fae79b6bcb67d43/dlib/dnn/trainer.h#L1032) `main_iteration_counter_at_last_disk_sync = main_iteration_counter;`, which means [`gradient_updates_since_last_sync`](https://github.com/davisking/dlib/blob/c2a9dee846cad7b20998ee0f3fae79b6bcb67d43/dlib/dnn/trainer.h#L1051) is always less than 30 (in my case, that is).

**Comment by davisking** [comment]
Oh, right, I forgot about that.  Yeah, don't call `get_net` during training
:)

**Comment by reunanen** [comment]
I want to do inference during training. Or at least stop the training for a while, do inference, then continue training.

Is it not necessary to call `get_net` to do inference?

**Comment by davisking** [comment]
Maybe there should be a version of get_net() that doesn't do the disk synchronization in that case.  Although really, I'm skeptical of the wisdom of stopping training to run inference.  I know lots of people do it.  But if there was some metric I wanted I would either compute it very infrequently so as to avoid disrupting training/slowing it down, or I would make a new loss layer that better captured whatever it was I was trying to optimize.

**Comment by reunanen** [comment]
I'm not trying to optimize anything outside of what the loss function represents. It's more that I really want to do actual inference as soon as possible. I understand the results won't be great just yet, but in my case it may still be better than no results at all. And yes, I do understand that it will slow down the training, at least a little.

**Comment by davisking** [comment]
What are you doing with the result of inference?

**Comment by reunanen** [comment]
Showing it to the user, for example. It's already quite helpful after a relatively small number of steps.

**Comment by reunanen** [comment]
Although this time the problem came from an automatic regression test, which is there to guarantee that any number of iterations above a certain minimum value produces a reasonable result for selected test input data. (This was not always the case, e.g. because of my misunderstanding that's now been documented in #870.)

**Comment by davisking** [comment]
Ok, but how often do you do it?  Not often enough for the syncing to disk to matter right?

**Comment by reunanen** [comment]
Probably not – eventually. Still more in a development & testing phase.

**Comment by reunanen** [comment]
Should [this](https://github.com/reunanen/dlib/commit/1b09482b8840a2c7f89211387db76b2ca9d452c0) work? Should I create a PR?

**Comment by davisking** [comment]
Yeah, that's a good idea.  In retrospect I should have made it that way in
the first place :)  So definitely send it as a PR :)
